### ISSUE TITLE
pgxc_clean should be invoked automatically?

### ISSUE THREAD

**Issue by DBInsight** [postgres-x2/postgres-x2#240]
when transaction partly commit (some node prepared and some nodes commit), we should invoke pgxc_clean manually to finish prepared transaction , if not, the prepared transaction held the lock and maybe block other operations, pgxc_clean should be invoked automatically? 

thoughts?

**Comment by koichi-szk** [comment]
Usually, it is reported as an error.   In such case, yes, pgxc_clean should be invoked.   pgxc_clean does not lock other transactions.   Pgxc_clean does not write to the database if no confining 2PC was found.

Regards;

**Comment by DBInsight** [comment]
yes, pgxc_clean has no problem. I mean when transaction prepared, although the transaction is ended , it does not release locks, select \* from pg_locks will see their held locks and their corresponding pid is 0, So if prepared transaction are not cleared in time, the prepared transaction can block other operations;

For example:
1. update xx; prepared , it will held RowExclusiveLock until it is committed or rollbacked;
if we do not invoke pgxc_clean manually, vacuum full will block.

thoughts?

**Comment by DBInsight** [comment]
yes, pgxc_clean has no problem. I mean when transaction prepared, although the transaction is ended , it does not release locks, select \* from pg_locks will see their held locks and their corresponding pid is 0, So if prepared transaction are not cleared in time, the prepared transaction can block other operations;

For example:
1. update xx; prepared , it will held RowExclusiveLock until it is committed or rollbacked;
if we do not invoke pgxc_clean manually, vacuum full will block.

thoughts?

**Comment by koichi-szk** [comment]
I suspect this could be common to PG itself.   If so, we should propose a patch to PG community and import it to X2.

**Comment by DBInsight** [comment]
Postgres has only explicit 2PC, but PGX2 use implicit 2PC, explicit 2PC should be cleared by user, but implicit 2PC used by PGX2 should be cleared automatically,  remaining implicit prepared transactions are caused by fault like DN power down or Network disconnect when CN sent down 'commit prepared'. should we reponsible for these remaining prepared transactions?

thoughts?

**Comment by koichi-szk** [comment]
This is what pgxc_clean does.   It will be a good idea to add pgxc_clean at each cluster startups.   With --status option, pgxc_ctl prints status of 2PC in the whole cluster.
